test/tall_format_test.dart

```diff
@@ -20,7 +21,118 @@ void main() async {
   await testDirectory('type', tall: true);
   await testDirectory('variable', tall: true);
 
-  // TODO(tall): The old formatter_test.dart has tests here for things like
-  // trailing newlines. Port those over to the new style once it supports all
-  // the syntax those tests rely on.
+  test('throws a FormatterException on failed parse', () {
+    var formatter = DartFormatter();
+    expect(() => formatter.format('wat?!'), throwsA(isA<FormatterException>()));
+  });
+
+  test('FormatterException.message() does not throw', () {
+    // This is a regression test for #358 where an error whose position is
+    // past the end of the source caused FormatterException to throw.
+    try {
+      DartFormatter().format('library');
+    } on FormatterException catch (err) {
+      var message = err.message();
+      expect(message, contains('Could not format'));
+    }
```

Comment:
[nit] Prefer to avoid `try/catch` in a test.

```suggestion
    expect(
        () => DartFormatter().format('library'),
        throwsA(isA<FormatterException>().having(
            (e) => e.message(), 'message', contains('Could not format'))));
```

Unless you specifically want the test to pass when no exception is thrown at all. In which case I think the following expresses the intent more explicitly.

```dart
    expect(
        () => DartFormatter().format('library'),
        anyOf(
            throwsA(isA<FormatterException>().having(
                (e) => e.message(), 'message', contains('Could not format'))),
            not(throwsA(anything))));
```

Reply:
Good idea. I'll apply this locally since I have other changes in progress.